### PR TITLE
[IMPL] Add core Vector3 functionality

### LINKED ISSUE #100: Expose core functionality on Vector3
#### Rationale

`Vector3` is currently a fairly opaque type. It should be update to expose a standard set of APIs such as are exposed in the DirectX Math library.

#### Proposed API

```csharp
public readonly struct Vector3
{
    public Vector3(float value);

    public static bool operator ==(Vector3 left, Vector3 right);
    public static bool operator !=(Vector3 left, Vector3 right);

    public static Vector3 operator +(Vector3 value);
    public static Vector3 operator -(Vector3 value);

    public static Vector3 operator +(Vector3 left, Vector3 right);
    public static Vector3 operator -(Vector3 left, Vector3 right);

    public static Vector3 operator *(Vector3 left, Vector3 right);
    public static Vector3 operator /(Vector3 left, Vector3 right);

    public static Vector3 operator *(Vector3 left, float right);
    public static Vector3 operator /(Vector3 left, float right);

    public static Vector3 Cross(Vector3 left, Vector3 right);
    public static float Dot(Vector3 left, Vector3 right);

    public static Vector3 Normalize(Vector3 value);

    public float Length { get; }
    public float LengthSquared { get; }
}
```

### PR DESCRIPTION
**Pull request by FiniteReality**
Fixes #100.

I've not vectorized these, and have instead only provided the software fallback, because I'm not sure I trust myself to correctly implement it. Plus, I believe there probably isn't much benefit to vectorizing *these* methods specifically, since they deal with small amounts of data - the overhead of shuffling data about would probably outweigh the performance benefits of vectorizing the code currently.

Maybe in the future we can add methods to take vectors in bulk to implement the vectorized versions of these algorithms.

Unit tests are in progress.

### CODE REVIEW

**Comment by FiniteReality** [comment]
I've currently omitted the single-parameter constructor since I cannot easily think of many scenarios where it would be clearer than `Vector3.One * value`.

**Comment by tannergooding** [comment]
> I've currently omitted the single-parameter constructor since I cannot easily think of many scenarios where it would be clearer than Vector3.One * value.

It's also for efficiency in the implementation. Once it's vectorized, it will forward to the implementation that does a broadcast, rather than having some other implementation.

File: sources/Core/Numerics/Vector3.cs [hunk 1]
```diff
@@ -48,6 +48,13 @@ public Vector3(float x, float y, float z)
         /// <summary>Gets the value of the z-dimension.</summary>
         public float Z => _z;
 
+        /// <summary>Gets the square-rooted length of the vector.</summary>
+        public float Length => MathF.Sqrt(LengthSquared);
+
+        /// <summary>Gets the squared length of the vector.</summary>
+        public float LengthSquared => Dot(this, this);
+
```
Comment (tannergooding) [inline sources/Core/Numerics/Vector3.cs:56]: One too many newlines:
```suggestion
```

File: sources/Core/Numerics/Vector3.cs [hunk 2]
```diff
@@ -70,6 +77,73 @@ public Vector3(float x, float y, float z)
...
+        /// <summary>Computes the cross product of two <see cref="Vector3"/> values.</summary>
+        /// <remarks>This method assumes both vectors <paramref name="left"/> and <paramref name="right"/> start at the origin.</remarks>
+        /// <param name="left">The first value to cross.</param>
+        /// <param name="right">The second value to cross.</param>
+        /// <returns>The cross product of <paramref name="left"/> and <paramref name="right"/></returns>
+        public static Vector3 Cross(Vector3 left, Vector3 right) => new Vector3(
+            (left.Y * right.Z) - (left.Z * right.Y),
+            (left.Z * right.X) - (left.X * right.Z),
+            (left.X * right.Y) - (left.Y * right.X));
```
Comment (tannergooding) [inline sources/Core/Numerics/Vector3.cs:134]: I personally prefer the following and find it slightly more parseable:
```suggestion
            (left.X * right.Y) - (left.Y * right.X)
        );
```

File: tests/Core/Numerics/Vector3Tests.cs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            Assert.True(+Vector3.Zero == Vector3.Zero);
+        }
+
+        /// <summary>Ensures that <see cref="Vector3.operator-(Vector3)"/> returns the negation of its input.</summary>
+        [Test]
+        public static void UnaryNegationReturnsNegative()
+        {
+            var minusOne = -new Vector3(1,1,1);
+
+            Assert.AreEqual(-1, minusOne.X);
```
Comment (tannergooding) [inline tests/Core/Numerics/Vector3Tests.cs:38]: The existing unit tests are using the NUnit constraint model, which is the recommended way to write tests: https://docs.nunit.org/articles/nunit/writing-tests/assertions/assertions.html

So this should probably be `Assert.That(-Vector3.One, Is.EqualTo(new Vector3(-1));`. Noting that it compares the entire vector at once, as I don't think doing the individual component comparisons provides much benefit (outside maybe the actual unit test for `.Equals`).

Comment (FiniteReality) [reply tests/Core/Numerics/Vector3Tests.cs:38]: I wrote the component comparisons separately to try and avoid implicitly testing multiple things; but I can use the operator if you desire.

Comment (tannergooding) [reply tests/Core/Numerics/Vector3Tests.cs:38]: I think you'll always end up testing multiple things as even the `X` property getter can have its own logic (and will once vectorization support goes in).

I'd rather keep the tests simpler and use things like the component comparisons in the actual `.Equals` test, where testing expected vs actual using the method you are testing makes less sense 😄

Comment (FiniteReality) [reply tests/Core/Numerics/Vector3Tests.cs:38]: True, true :sweat_smile:

**Review by tannergooding — APPROVED** [review]

**Pull request merged**